protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
+Transaction $tx$ is added to $p.lcommitted$ when step (2).
+It is removed when (4) takes place.
+Step (6) adds the transaction to the mempool.
+The transaction is removed from the mempool upon (7) and (8).
+
+Consider some correct process $p$.
+Below, we show in order that all the invariants of the mempool are maintained at $p$:  
+
+- INV1.
+Clear from the code base, as the consensus input corresponds to a call to `Execution:CreateProposalBlock` which itself is calling `ClistMempool:ReapMaxBytesMaxGas`, returning a subset of $p.mempool$.
+
+- INV2a.
+If $tx$ is in $p.lcommitted$, it was added due to step (2).
+This step occurs only if the transaction was committed at some height.
+Furthermore, when this happens step (7) is taken.
+It follows, that the transaction is removed from the mempool, as required.
+
+- INV3a.
+Consider some height $h$ for which $tx$ is in the mempool and $p.valid[tx]>0$.
+For any transaction $tx$, this counter is set to $0$ initially.
+Hence, $p.valid[tx]$ was incremented at some height $h' \leq h$.
+Consider the point in time $t'$, where p moves to height $h$.
+If $p.valid[tx]$ equals 0 at that time, then it was incremented at height h.
+From (6), it is necessarily valid at height h, as requied.
+Otherwise, the transaction is rechecked at the current (new) height h with step (8).
+Because $p.valid[tx]$ is not reset, it follows that the application considers it valid at h 
+
+- INV4.
+Consider that a transaction $tx$ enters the mempool at $p$.
+If $tx$ leaves the mempool, step (7) or (8) happens.
+In the former case, $tx$ is committed as required.
+In the latter, $tx$ is re-checked and invalid at some height $h$.
+Re-checking a transaction happens only once per height, say at time $t$ for transaction $tx$ and height $h$.
+When this happens at time $t$ (line 474), the transaction is also removed from the cache (line 477) due to step (3).
+By assumption, step (1) is eventually re-executed for transaction $tx$ after time $t$.
+Because $tx$ is no more in the cache, it is added again after time $t$.
+Now, assume that transaction $tx$ is not invalid infinitely often at process $p$.
+Hencce after some time, it is always valid at $p$.
+Observe that this is also eventually the case at all the correct processes in the system.
+Name $t'$ the moment in time when this occurs.
+By assumption, there are only a bounded amount of transactions submitted in the system.
+After time $max(t,t')$, $tx$ is eventually submitted at some height at which it is valid and committed.
+
+*A remark.* 
+The reader might observe that in the above reasoning we do not use step (5).
+This comes from the fact that stricly speaking this step is not necessary for the mempool invariants to hold.
+One could consider a stronger variation of INV4 in which a transaction is eventually either 
+forever invalid,
+or at some height both valid and committed.
+This variation would necessitate step (5) to oust a committed yet invalid transaction from the cache.
```

Comment:
Probably one if the AIs coming from this spec is what to do with step (5). Potentially a subject for a comm call, to see what users think, and whether removing it would break existing Apps (I don't think so, but that's jut my view).

Reply:
Indeed, this is an interesting discussion. 

Step (5) comes from the fact that CometBFT allows an application to report as invalid some of the committed transactions. To some extend, this is not usual. From what I see, this is not the case in [Ethereum](https://docs.alchemy.com/docs/how-to-cancel-a-private-transaction-on-ethereum), nor in the [RedBelly](https://csrg.redbellyblockchain.io/doc/ConsensusRedBellyBlockchain.pdf) blockchain, nor in [this](https://arxiv.org/pdf/1805.08429.pdf) attempt to formalize Tendermint.

Nonetheless, as we discussed, allowing this permits to hide latency as the mempool may contain transactions that depend one from another (aka., pipelining). Due to byzantine nodes, such a dependency might be broken at commit time.

Reply:
From a user's perspective, (a) I would imagine myself submitting tx1, and then waiting for it to appear in a block. If the tx1 appears in a block as failed, I (as a user) would be expected to re-issue the tx again (re-try).

The alternative would be (b) for me (the user) to think: as tx1 failed in block B, I don't need to do anything: it will re-try by itself.

I find (a) much simpler.

Reply:
I didn't get the full point here. The transaction is committed but invalid, so it is removed from the cache, ok. But it is also removed from the mempool, right? If so, how can it be included in a future block?